README.md
# Python Terminal SuperMario


This **README** file contains :
 1. Team information
 2. Requirements
 3. How to run
 4. File structure
 5. Requirements
 6. Video link - https://youtu.be/-1zx1T2olzc

----------


Team Information
-------------

- **Prajwal Krishna**
- **Yash Chaurasia**
- **Aradhya Tongia**
- **Pratyush Kumar**

----------
Requirements:
--------------------
- Python3
- Rest of the requirements are mentioned in **requirements.txt** file
- Active internet connection
---------------------
## Running the program

- First, install all the requirements:
	- `pip install -r requirements.txt`
- Change the folder
	- `cd Website`
- Type the following command
    - `python3 flaskblog.py`
- Open a browser and paste the following link
    - `http://127.0.0.1:5000/`

------------------------------


## Running other parts of project:
##### Just run python scripts as
`python3 <script name>`

-------
## File Structure

 * [twitterScraper.py](./twitterScraper.py) -   Twitter Scrapper file which helps to populate the dataset
 * [dataCleaning.py](./dataCleaning.py) - File which helps to preprocessing and cleaning the dataset
 * [LogisticRegression.ipynb](./LogisticRegression.ipynb) - Jupyter notebook which runs logistic regression
 * [Dataset](./Dataset) - Stores 200 recent tweets of various persons
 * [Features](./Features) - Stores JSON response of IBM Watson Personality Insights for each person in database
 * [Friends](./Friends) - List of friends amongst persons in the database
 * [auxilary_function.py](./Website/auxilary_function.py) - Useful functions to handle new users from webapp
 * [logisticRegression.pkl](./logisticRegression.pkl) - Pickle file which stores trained logistic regression parameters
 * [records.csv](./records.csv) - Records of people and friendship
 * [Undertaking_form.pdf](./Undertaking_form.pdf) - Undertaking form
 * [Presentation.pptx](./Presentation.pptx) - Powerpoint presentation
 * [Idea.pdf](./Idea.pdf) - Idea Documentation
 * [requirements.txt](./requirements.txt) - File containing all python dependencies.
 * [README.md](./README.md) - This file

**Note:** The development was done on a UNIX environment

## Video Link
-----------------
Link to video - (https://youtu.be/-1zx1T2olzc)
_______________

#### Prajwal Krishna Maitin


Dataset/HISTORY.txt
⚡️ “The Moon Landing: 50 Years Later” #Apollo50th 
https://t.co/Wx54gezMhh
On #ThisDayinHistory 1881, Sioux leader Sitting Bull surrenders to the U.S. Army, which promises amnesty for him and his followers. https://t.co/Qw5MLe1XW7
Journey into the darkness, the mysterious, the deadly...The #UnXplained. https://t.co/hAU0KJ6dw0
RT @AncientAliens: Live tweet with @WilliamShatner during the premiere of #UnXplained at 10/9c. https://t.co/c969kX8uhy
Join @WilliamShatner on a journey into the #UnXplained. Be warned. It's not for the faint of heart. https://t.co/jazdxRWYLg
Explore the state-sponsored drug use that helped build and bring down Nazi Germany. Nazis On Drugs: Hitler And The Blitzkrieg premieres Sunday night at 9/8c on History. https://t.co/I76O4wlBFg
On #ThisDayinHistory 1848, at the Wesleyan Chapel in Seneca Falls, New York, a woman’s rights convention—the first of its kind ever held in the United States—convenes. https://t.co/rV7PT1LPid
One small snip for man, one giant con for the barber. Neil Armstrong's barber did WHAT with his hair?! https://t.co/YKMMqSYJlj
After only nine days as the monarch of England, Lady Jane Grey is deposed in favor of her cousin Mary on #ThisDayinHistory 1553. The 15-year-old Lady Jane had only reluctantly agreed to be put on the throne and the decision would result in her execution.https://t.co/OYe7YvsxXF
Friday is the strongest day. It's where the weak-ends. 💪
Don't miss a new #StrongestMan in History, Wednesday at 10/9c. https://t.co/BgJH4YZ2LR
"We're in trouble," said Commander Tom Stafford. https://t.co/BFsW55x4hh
🚨WARNING🚨
 What you're about to watch may be disturbing to some viewers. #AloneShow https://t.co/6SCXBCxbB7
If last night's #AxMen had you on the edge of your big rig, hold onto your hard hats and see what's in store for next Thursday night! https://t.co/wywkXDhUfc
The Etienne's are one of #AxMen's newest families to join the operation. As they say, the family that stays together, works together. And you don't want to mess with Big Mama. Ax Men is tonight at 9/8c on History. https://t.co/fqxISsorgz
On #ThisDayinHistory 1936, the Spanish Civil War begins as a revolt by right-wing Spanish military officers. https://t.co/ynfauvhJR6
The third-longest running member of the court changed his mind on just about every hot button issue of the day during his term. https://t.co/b6ofCv56Cf
In fact, the ships that were raided by the Sons of Liberty were built and owned by Americans. https://t.co/HJaUDnUkfK
Back to the grind. Don't forget your tools. #AxMen is back tonight at 9/8c on History. https://t.co/DGv2WFQyWh
The spaceship flight computer had circuitry that was hand-woven. Each wire representing a 1 or a 0 in the computer program had to be positioned with absolute precision, by a person, using a needle, and wire instead of thread. https://t.co/bZ2DKE8aii
Did Hershey chocolate help America win WWII? Feed your curiosity, beginning August 11 at 9/8c. #FoodThatBuiltAmerica https://t.co/c3MrMj6Fv6
One of the earliest human computers still works at JPL. NASA’s longest-serving female employee, Sue Finley was originally hired in 1958 to work on trajectory computations for rocket launches. #apollo50th https://t.co/ktaMUIQXOv
On #ThisDayinHistory 50 years ago, then Senator Edward “Ted” Kennedy of Massachusetts drives an Oldsmobile off a wooden bridge into a tide-swept pond. Kennedy escaped the submerged car, but his passenger, 28-year-old Mary Jo Kopechne, did not.https://t.co/22xo8Cxnis
Make sure you're wearing your hard hat when #AxMen takes to the land tomorrow night at 9/8c on History. https://t.co/QjYJlp1jfr
The spaceship flight computer had circuitry that was hand-woven. Each wire representing a 1 or a 0 in the computer program had to be positioned with absolute precision, by a person, using a needle, and wire instead of thread. https://t.co/0cJA1mKKnq
@DavidHodorowski @eddiehallWSM https://t.co/tvfl4auFeF
@RyanGeeksOut @eddiehallWSM https://t.co/nNCZfBw85T
@CurseOfOak @eddiehallWSM https://t.co/QecUeLURIH
@americanpickers @eddiehallWSM @AmericanPicker @FritzPicker https://t.co/cPfWlUAUMQ
@HistoryVikings @eddiehallWSM #Vikings https://t.co/fDZdpDh8FE
@amg6578 @eddiehallWSM https://t.co/akA7EQB4oD
@TheFibDoctor @eddiehallWSM https://t.co/dLmL1CkhoT
@mad2988 @eddiehallWSM @shawstrength https://t.co/f9Qf2I1fLS
@LittleWomenTV @eddiehallWSM #LittleWomenATL https://t.co/kteynCjOSX
@MapleWicket @eddiehallWSM https://t.co/gdMGBqkxf6
@CandiceHere_ @eddiehallWSM @GillianA @GillianA https://t.co/nzQHLZeMVM
@shawnhollenbach @eddiehallWSM https://t.co/8ZziWXY4Er
@TheDVK @eddiehallWSM https://t.co/OIqV1f6xY3
@KristinaWagner0 @eddiehallWSM https://t.co/f5EpRzmnfm
@mintbury @eddiehallWSM https://t.co/HqcFhFOLFE
@gt9054a @eddiehallWSM https://t.co/d0wz3NBMKk
@josephcisar @pawnstars @eddiehallWSM https://t.co/pf54Um5cop
@SusanKulstad @eddiehallWSM https://t.co/by6uC7Bii4
@eploven @eddiehallWSM https://t.co/EX9wfUTov3
@Primo_Dulce @eddiehallWSM https://t.co/MMLdKbpgYh
The time has come. Tell us what you want @eddiehallWSM to lift below! https://t.co/BYJ0SFwg0J
The Apollo program had survived three administrations, one major accident that killed three astronauts, and the tumultuous 1960s to realize the dream of a President who wasn’t alive to see it. #apollo50th https://t.co/ZE5B2Ndufj
On #ThisDayinHistory 75 years ago an ammunition ship explodes in Port Chicago, California, killing 332. Nearly 2/3 of those killed were African-American enlisted men in the Navy–15% percent of all African-Americans killed during World War II. https://t.co/s75ZFqEOeX
One mistake could end someone's career...or worse. 😳
Tonight, see how far these guys are willing to go to prove they are The Strongest Men in History. https://t.co/aSP88CuZey
Make @eddiehallWSM raise the bar. What do YOU want to see him lift? Answer below and we’ll surprise you with some fun lifts! 👇😁 https://t.co/dhSIfvOWvZ

Website/func.py
import tweepy
import requests
import json
import os.path
import numpy as np
import pandas as pd
import pickle

consumer_key = 'Kt6WNY1DPSBLPkrFzQoE0TV6h'
consumer_secret = 'UUTUOUgIZ0uBijCWay4Y1j6Spb0JlYBFvngB4a1WeC1KK9kQmg'
access_token = '2951964404-Thlks6h0r7mboOf36Yk3dp1YCKgHYho0QdY7uCD'
access_token_secret = 'kBInoOMXY362LkAvRq87geWpqtdIxL93jOxviICVmFm1k'
ibmAuthoziation = 'mYMCHtOw3tfKY2ulgkbR2qv_Jc1UP26w3bN5Gf4zXxFP'
ibmBaseUrl = 'https://gateway-lon.watsonplatform.net/personality-insights/api'
ibmPersonalityUrl = ibmBaseUrl + '/v3/profile?version=2017-10-13'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
# Calling api
api = tweepy.API(auth)

def get_tweets(username):
    if(os.path.isfile('./Dataset/' + username + '.txt')):
        return
    auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_token_secret)
    # # Calling api
    api = tweepy.API(auth)
    # 200 tweets to be extracted
    number_of_tweets=2000
    tweets = api.user_timeline(screen_name=username,count=number_of_tweets,tweet_mode="extended")

    f = open("./Dataset/" + username + ".txt", "w")
    for tweet in tweets:
        f.write(tweet.full_text + '\n')
    f.close()

def get_traits(username):
    if(os.path.isfile('./Features/' + username + '.json')):
        return
    path = "./Dataset/" + username + ".txt"
    data = open(path, 'rb').read()
    res = requests.post(
            ibmPersonalityUrl,
            auth=('apikey',ibmAuthoziation),
            headers={
                'Content-Type':'text/plain',
                'Accept':'application/json'
                },
            data = data
            )
    os.remove('./Dataset/' + username + '.txt')
    return res.json()

def cleanData(jsonResponse,user):
    try:
        big5 = jsonResponse['personality']
        user = { 'Name' : user}
        for i in big5:
            user[i['name']]=i['percentile']*100
        return user
    except:
        return False

def run_logistic_regression(data):
    with open('../logisticRegression.pkl', 'rb') as f:
        logreg = pickle.load(f)
    train = pd.read_csv("./users.csv")
    train.sample(frac=1)
    X = train.drop(columns=["Name"],axis=1)
    name_array = list(train["Name"])
    X['User Openness'] = data['Openness']
    X['User Conscientiousness'] = data['Conscientiousness']
    X['User Extraversion'] = data['Extraversion']
    X['User Agreeableness'] = data['Agreeableness']
    X['User Emotional range'] = data['Emotional range']
    X = X.rename(columns={
        'Openness':'Friend Openness',
        'Conscientiousness':'Friend Conscientiousness',
        'Extraversion':'Friend Extraversion',
        'Agreeableness':'Friend Agreeableness',
        'Emotional range':'Friend Emotional range'
        })
    X = X[['User Openness','User Conscientiousness','User Extraversion','User Agreeableness','User Emotional range',
        'Friend Openness','Friend Conscientiousness','Friend Extraversion','Friend Agreeableness','Friend Emotional range']]
    H = logreg.decision_function(X)
    arr = np.array(H)
    a = arr.argsort()[-10:][::-1]
    ans = []
    for i in a:
        ans.append(name_array[i])
    return ans

def test(user):
    get_tweets(user)
    personality = get_traits(user)
    data = cleanData(personality,user)
    if(data):
        return run_logistic_regression(data)


dataCleaning.py
import json
import os
import csv
users = []
for filename in os.listdir('./Features/'):
    with open('./Features/' + filename) as json_file:
        data = json.load(json_file)
        try:
            big5 = data['personality']
            user = { 'Name' : filename[:-5]}
            for i in big5:
                user[i['name']]=i['percentile']*100
            # needs = data['needs']
            # for need in needs:
            #     user[need['name']]=i['percentile']*100
            users.append(user)
        except:
            pass
print(len(users))
keys = users[0].keys()
with open('users.csv', 'w') as output_file:
    dict_writer = csv.DictWriter(output_file, keys)
    dict_writer.writeheader()
    dict_writer.writerows(users)

records = []
# Making pairs
for user in users:
    username = user['Name']
    friends = []
    count = 25
    friends.append(username)
    if(os.path.isfile('./Friends/' + username + '.json')):
        with open('./Friends/' + username + '.json') as json_file:
            data = json.load(json_file)
            for friend in data:
                friends.append(friend)
    if(len(friends)<3):
        continue
    for id in users:
        record = {}
        for field in user:
            record['User ' + field] = user[field]
        for field in id:
            record['Friend ' + field] = id[field]
        if id['Name'] in friends:
            record['Friendship Status'] = 1
            records.append(record)
        else:
            record['Friendship Status'] = 0
            if(count):
                records.append(record)
                count -= 1
print(len(records))
keys = records[0].keys()
with open('records.csv', 'w') as output_file:
    dict_writer = csv.DictWriter(output_file, keys)
    dict_writer.writeheader()
    dict_writer.writerows(records)


twitterScraper.py
import tweepy
import requests
import json
import os.path

consumer_key = 'Kt6WNY1DPSBLPkrFzQoE0TV6h'
consumer_secret = 'UUTUOUgIZ0uBijCWay4Y1j6Spb0JlYBFvngB4a1WeC1KK9kQmg'
access_token = '2951964404-Thlks6h0r7mboOf36Yk3dp1YCKgHYho0QdY7uCD'
access_token_secret = 'kBInoOMXY362LkAvRq87geWpqtdIxL93jOxviICVmFm1k'
ibmAuthoziation = 'mYMCHtOw3tfKY2ulgkbR2qv_Jc1UP26w3bN5Gf4zXxFP'
ibmBaseUrl = 'https://gateway-lon.watsonplatform.net/personality-insights/api'
ibmPersonalityUrl = ibmBaseUrl + '/v3/profile?version=2017-10-13'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
# Calling api
api = tweepy.API(auth)

def get_tweets(username):
    if(os.path.isfile('./Dataset/' + username + '.txt')):
        return
    auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_token_secret)
    # # Calling api
    api = tweepy.API(auth)
    # 200 tweets to be extracted
    number_of_tweets=2000
    tweets = api.user_timeline(screen_name=username,count=number_of_tweets,tweet_mode="extended")

    f = open("./Dataset/" + username + ".txt", "w")
    for tweet in tweets:
        f.write(tweet.full_text + '\n')
    f.close()

def get_traits(username):
    if(os.path.isfile('./Features/' + username + '.json')):
        return
    path = "./Dataset/" + username + ".txt"
    data = open(path, 'rb').read()
    res = requests.post(
        ibmPersonalityUrl,
        auth=('apikey',ibmAuthoziation),
        headers={
            'Content-Type':'text/plain',
            'Accept':'application/json'
        },
        data = data
        )
    # print(res.text)

    response = res.json()
    # print(res.json())
    with open('./Features/' + username + '.json', 'w') as json_file:
        json.dump(response, json_file)
    # print(res)

def get_friends(username, depth=0):
    auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_token_secret)
    # Calling api
    api = tweepy.API(auth)

    friends = api.friends_ids(screen_name=username)
    used_friends = set()
    if(len(friends)>10):
        friends = friends[2:10]
    for friend in friends:
        user = api.get_user(id=friend)
        print('cur ' + user.screen_name)
        get_tweets(user.screen_name)
        get_traits(user.screen_name)
        used_friends.add(user.screen_name)
        if(depth<2):
            get_friends(user.screen_name,depth+1)
    if(os.path.isfile('./Friends/' + username + '.json')):
        with open('./Friends/' + username + '.json') as json_file:
            data = json.load(json_file)
            for i in data:
                used_friends.add(i)
    print(list(used_friends))
    with open('./Friends/' + username + '.json', 'w') as json_file:
        json.dump(list(used_friends), json_file)

if __name__ == '__main__':
    get_friends("y5yash",1)


Website/auxilary_function.py
import tweepy
import requests
import json
import os.path
import numpy as np
import pandas as pd
import pickle

consumer_key = 'Kt6WNY1DPSBLPkrFzQoE0TV6h'
consumer_secret = 'UUTUOUgIZ0uBijCWay4Y1j6Spb0JlYBFvngB4a1WeC1KK9kQmg'
access_token = '2951964404-Thlks6h0r7mboOf36Yk3dp1YCKgHYho0QdY7uCD'
access_token_secret = 'kBInoOMXY362LkAvRq87geWpqtdIxL93jOxviICVmFm1k'
ibmAuthoziation = 'mYMCHtOw3tfKY2ulgkbR2qv_Jc1UP26w3bN5Gf4zXxFP'
ibmBaseUrl = 'https://gateway-lon.watsonplatform.net/personality-insights/api'
ibmPersonalityUrl = ibmBaseUrl + '/v3/profile?version=2017-10-13'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
# Calling api
api = tweepy.API(auth)

def get_tweets(username):
    if(os.path.isfile('./Dataset/' + username + '.txt')):
        return
    auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_token_secret)
    # # Calling api
    api = tweepy.API(auth)
    # 200 tweets to be extracted
    number_of_tweets=2000
    tweets = api.user_timeline(screen_name=username,count=number_of_tweets,tweet_mode="extended")

    f = open("./Dataset/" + username + ".txt", "w")
    for tweet in tweets:
        f.write(tweet.full_text + '\n')
    f.close()

def get_traits(username):
    if(os.path.isfile('./Features/' + username + '.json')):
        return
    path = "./Dataset/" + username + ".txt"
    data = open(path, 'rb').read()
    res = requests.post(
            ibmPersonalityUrl,
            auth=('apikey',ibmAuthoziation),
            headers={
                'Content-Type':'text/plain',
                'Accept':'application/json'
                },
            data = data
            )
    os.remove('./Dataset/' + username + '.txt')
    return res.json()

def cleanData(jsonResponse,user):
    try:
        big5 = jsonResponse['personality']
        user = { 'Name' : user}
        for i in big5:
            user[i['name']]=i['percentile']*100
        return user
    except:
        return False

def run_logistic_regression(data):
    with open('../logisticRegression.pkl', 'rb') as f:
        logreg = pickle.load(f)
    train = pd.read_csv("./users.csv")
    train.sample(frac=1)
    X = train.drop(columns=["Name"],axis=1)
    name_array = list(train["Name"])
    X['User Openness'] = data['Openness']
    X['User Conscientiousness'] = data['Conscientiousness']
    X['User Extraversion'] = data['Extraversion']
    X['User Agreeableness'] = data['Agreeableness']
    X['User Emotional range'] = data['Emotional range']
    X = X.rename(columns={
        'Openness':'Friend Openness',
        'Conscientiousness':'Friend Conscientiousness',
        'Extraversion':'Friend Extraversion',
        'Agreeableness':'Friend Agreeableness',
        'Emotional range':'Friend Emotional range'
        })
    X = X[['User Openness','User Conscientiousness','User Extraversion','User Agreeableness','User Emotional range',
        'Friend Openness','Friend Conscientiousness','Friend Extraversion','Friend Agreeableness','Friend Emotional range']]
    H = logreg.decision_function(X)
    arr = np.array(H)
    a = arr.argsort()[-10:][::-1]
    ans = []
    for i in a:
        ans.append(name_array[i])
    return ans

def utilityFunction(user):
    get_tweets(user)
    personality = get_traits(user)
    data = cleanData(personality,user)
    if(data):
        return run_logistic_regression(data)
